### PR TITLE
lp-373 : dto validation on post endpoint

### PR DESCRIPTION
**Pull request by lduranteau**
### JIRA link
https://companieshouse.atlassian.net/browse/LP-373

### Change description
add dto validation on post request

### Work checklist

* [ ] Bug fix
* [x] New feature
* [ ] Breaking change
* [ ] Infrastructure change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [x] I have added/updated functional tests where appropriate
* [ ] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__
### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java [hunk 1]
```diff
@@ -32,7 +39,7 @@ public void setPartnershipName(String partnershipName) {
     }
 
     public String getNameEnding() {
-        return nameEnding.getDescription();
+        return nameEnding != null ? nameEnding.getDescription() : null;
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:42]: Wondering why this implementation change was needed? Was it due to a NullPointerException happening on line 18 of the `NameSizeValidator` class maybe? If so, might be slightly cleaner to deal with that null value in that validator class rather than change this.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java [hunk 2]
```diff
@@ -2,21 +2,28 @@
...
+import uk.gov.companieshouse.limitedpartnershipsapi.model.dto.validator.NameSize;
 
 import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
 
+@NameSize
 public class DataDto {
     @JsonInclude(NON_NULL)
     @JsonProperty("partnership_name")
+    @Size(min = 1, message = "partnership_name must be greater than 1")
+    @Size(max = 160, message = "partnership_name must be less than 160")
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:18]: Could we define the max length of 160 somewhere as a constant (maybe within this class - e.g. `PARTNERSHIP_NAME_MAX_LENGTH`) and then reference it in all those places where 160 is used (four in the application code I think).

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 1 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:42]: This has nothing to do with NameSizeValidator
This is just because if nameEnding is null an exception is thrown as it tries to use the PartnershipNameEnding method

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 2 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:18]: Where do you see 4?

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 2 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:18]: The 160 maximum includes the `name_ending` field.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 2 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java]: @dlloyd1-ch  there is a custom validator for that

Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java]: constants added

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 1 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:45]: 👍 is it possible to add a unit-test that exercises the scenario when name ending is null?

(when I revert this change all the current tests still pass)

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java [hunk 3]
```diff
@@ -2,21 +2,31 @@
...
+import jakarta.validation.constraints.Size;
 import uk.gov.companieshouse.limitedpartnershipsapi.model.PartnershipNameEnding;
 import uk.gov.companieshouse.limitedpartnershipsapi.model.PartnershipType;
+import uk.gov.companieshouse.limitedpartnershipsapi.model.dto.validator.NameSize;
 
 import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
 
+@NameSize
 public class DataDto {
+    public static final int MIN_SIZE = 1;
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:15]: I would rename these constants to `PARTNERSHIP_NAME_MIN_SIZE` and `PARTNERSHIP_NAME_MAX_SIZE` as there are other fields in this DTO that the sizes could refer to, e.g. `email`.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/model/LimitedPartnershipSubmissionDtoValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        assertFalse(violations.isEmpty());
+        assertThat(violations).hasSize(3);
+        assertThat(violations)
+                .extracting(ConstraintViolation::getMessage)
+                .containsExactlyInAnyOrder(
+                        String.format("Max length 'data.partnership_name + data.name_ending' is %s characters", DataDto.MAX_SIZE),
+                        String.format("partnership_name must be less than %s", DataDto.MAX_SIZE),
+                        "must be a well-formed email address");
+    }
+
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/model/LimitedPartnershipSubmissionDtoValidationTest.java:70]: Could we add a test in here to ensure that email is also validated?

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 3 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:15]: I use it to annotate the PartnershipName but in reality it is PartnershipName + NameEnding (custom validator)
Maybe NameSize_ is better

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/model/LimitedPartnershipSubmissionDtoValidationTest.java (hunk 4 above)
Comment (lduranteau) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/model/LimitedPartnershipSubmissionDtoValidationTest.java:70]: have a look to line 68

Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/model/LimitedPartnershipSubmissionDtoValidationTest.java:70]: Sorry, missed that. All good.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java (hunk 3 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/DataDto.java:15]: 👍 yes, so maybe just `NAME_MIN_SIZE` and `NAME_MAX_SIZE` then.

**Review by mwestacott — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/validator/NameSizeValidator.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.ConstraintValidatorContext;
+import uk.gov.companieshouse.limitedpartnershipsapi.model.dto.DataDto;
+
+public class NameSizeValidator implements ConstraintValidator<NameSize, DataDto> {
+    public boolean isValid(DataDto object, ConstraintValidatorContext context) {
+        if (!(object instanceof DataDto)) {
+            throw new IllegalArgumentException("@NameSize only applies to DataDto object");
+        }
+
+        return object.getPartnershipName().length() + object.getNameEnding().length() <= DataDto.NAME_MAX_SIZE;
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/model/dto/validator/NameSizeValidator.java:13]: Also need to allow for a space between the partnership_name and the name_ending.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by dlloyd1-ch — APPROVED** [review]
